apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,46 @@
+import asyncHandler = require('express-async-handler');
+import * as sqldb from '@prairielearn/postgres';
+import { IdSchema } from '../lib/db-types';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  const user_session_id = await sqldb.queryOptionalRow(
+    sql.select_session_id,
+    { user_session_id: req.session.id },
+    IdSchema,
+  );
+
+  const params = {
+    ip_address: req.ip,
+    user_id: res.locals.authn_user.user_id,
+    user_session_id: user_session_id,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
+  };
+  let client_fingerprint_id = await sqldb.queryOptionalRow(
+    sql.select_client_fingerprint,
+    params,
+    IdSchema,
+  );
+
+  if (!client_fingerprint_id) {
+    client_fingerprint_id = await sqldb.queryRow(sql.insert_client_fingerprint, params, IdSchema);
+  }
+  if (res.locals.assessment_instance) {
+    if (
+      res.locals.assessment_instance.last_client_fingerprint_id &&
+      IdSchema.parse(res.locals.assessment_instance?.last_client_fingerprint_id) !==
+        client_fingerprint_id
```

Comment:
I have now updated to use `!idsEqual` and I have removed the check for `res.locals.assessment_instance.last_client_fingerprint_id`. 

It does look like client fingerprint change count is now incrementing from `0` to `1` on the first set of `last_client_fingerprint_id`. However, I'm not sure how I can avoid incrementing the change count purely in the sql query. Do you have any suggestions on how to accomplish this?

Reply:
I believe you can achieve this by adjusting the query as follows:

```sql
UPDATE assessment_instances AS ai
SET
  last_client_fingerprint_id = $client_fingerprint_id,
  client_fingerprint_id_change_count = CASE
    WHEN ai.last_client_fingerprint_id IS NULL THEN 0
    ELSE client_fingerprint_id_change_count + 1
  END
WHERE ...
```

This says "if `last_client_fingerprint_id` is NULL, set the change count to 0, otherwise increment it". I'd recommend testing this first, but I believe it will work.

Reply:
Thanks! I have updated the SQL query to this in commit [c4eb708](https://github.com/PrairieLearn/PrairieLearn/pull/8798/commits/c4eb70857d0fce8cd75fdefd6e3694c0850b25b7). I have tested and this works as intended.